pkg/porter/reconcile_test.go

```diff
@@ -0,0 +1,174 @@
+package porter
+
+import (
+	"path/filepath"
+	"testing"
+
+	"get.porter.sh/porter/pkg/claims"
+	"get.porter.sh/porter/pkg/cnab"
+	portercontext "get.porter.sh/porter/pkg/context"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPorter_IsInstallationInSync(t *testing.T) {
+	cxt := portercontext.New()
+	bun, err := cnab.LoadBundle(cxt, filepath.Join("testdata/bundle.json"))
+	require.NoError(t, err)
+
+	t.Run("inactive not yet installed", func(t *testing.T) {
+		p := NewTestPorter(t)
+		defer p.Teardown()
+
+		i := claims.Installation{
+			Active: false,
+		}
+		insync, err := p.IsInstallationInSync(p.RootContext, i, nil, NewInstallOptions())
+		require.NoError(t, err)
+		assert.True(t, insync)
+		assert.Contains(t, p.TestConfig.TestContext.GetError(), "Ignoring because the installation is inactive")
+	})
+
+	t.Run("active not yet installed", func(t *testing.T) {
+		p := NewTestPorter(t)
+		defer p.Teardown()
+
+		i := claims.Installation{
+			Active: true,
+		}
+		insync, err := p.IsInstallationInSync(p.RootContext, i, nil, NewInstallOptions())
+		require.NoError(t, err)
+		assert.False(t, insync)
+		assert.Contains(t, p.TestConfig.TestContext.GetError(), "Triggering because the installation has not completed successfully yet")
+	})
+
+	t.Run("active installed - no changes", func(t *testing.T) {
+		p := NewTestPorter(t)
+		defer p.Teardown()
+
+		i := claims.Installation{
+			Active: true,
+			Status: claims.InstallationStatus{
+				Installed: &now,
+			},
+		}
+		run := claims.Run{
+			// Use the default values from the bundle.json so that we don't trigger reconciliation
+			Parameters: map[string]interface{}{
+				"my-second-param": "spring-music-demo",
+			},
+		}
+		upgradeOpts := NewUpgradeOptions()
+		upgradeOpts.bundleRef = &cnab.BundleReference{Definition: bun}
+		insync, err := p.IsInstallationInSync(p.RootContext, i, &run, upgradeOpts)
+		require.NoError(t, err)
+		assert.True(t, insync)
+		// Nothing is printed out in this case, the calling function will print "up-to-date" for us
```

Comment:
I wonder if it's still a good idea to assert this behavior so we know when it changes in the future

Reply:
Other tests check for the "up-to-date" message that is printed. We can't check for it in this unit test because the function that prints it isn't called.

Here is my sad attempt at visualizing the call tree and who prints what
```
ApplyInstallation()
  -> IsInstallationInSync()
          prints if it should be triggered
  performs extra logic and prints if the installation is up to date or what action will be executed
```

Reply:
I see. That make sense. What about asserting that nothing is printed out, so we can confirm this expected behavior in the test?
```go
assert.Contains(t, p.TestConfig.TestContext.GetError(), "")
```

Reply:
When I went to add that check I discovered that other functions end up printing to stderr. I don't think it's worth asserting that nothing is printed in this case, it would just make the test more fragile.

I was checking for the existence of the "triggering" and "up-to-date" because they help give the user context into what's going on, and why things happened or didn't. But checking for the lack of printing something is something that is much easier to break when we make changes, like adding more log statements elsewhere. And elsewhere in our tests we do check for "up-to-date", so I feel like we have this code path covered pretty well already.